core/modules/ccontrol/testAntlr4GeneratedIR.cc

```diff
@@ -65,60 +66,52 @@ using namespace lsst::qserv;
- * @param v The last object to be pushed.
- */
-template <typename T>
-void pusher(vector<T>& vec, T v) {
-    vec.push_back(v);
-}
-
 
 /**
  * @brief Pusher is a set of recursive variadic functions to receive a variable number of arguments and push
- *        them onto a vector.
+ *        them into a container that has a push_back function.
  *
- *        This form is used when explicitly decalring the template types when there is exactly one arg.
+ *        This form of `pusher` is the last to get called, for the single remaining element.
  *
- * @tparam T The type contained by the vector.
- * @tparam U The type of object being pushed.
- * @param vec The vector to push onto.
- * @param v The last object to be pushed.
+ * @tparam Container The type of container
+ * @tparam Types The template parameters the container accepts (e.g. for a vector it will be the element type
+ *               and the allocator type)
+ * @tparam T The type of the object to push into the container.
+ * @tparam TArgs The type of the remaining objects to push onto the vector (this should match T)
+ * @param container The container to push objects into.
+ * @param first The object to push into the container.
  */
-template <typename T, typename U>
-void pusher(vector<T>& vec, U v) {
-    vec.push_back(v);
+template <template <typename... Args> class Container, typename... Types, typename T, typename... TArgs>
+void pusher(Container<Types...>& container, T first) {
+    container.push_back(first);
 }
 
 
 /**
  * @brief Pusher is a set of recursive variadic functions to receive a variable number of arguments and push
- *        them onto a vector.
+ *        them into a container that has a push_back function.
  *
- * @tparam T The type contained by the vector.
- * @tparam Args The variadic list of objects to push onto the vector.
- * @param vec The vector to push onto.
- * @param first The object to push onto. This will be each next object until there is one left, then the
- *              other form of `pusher` will be called.
- * @param args The rest of the object to be pushed onto the vector.
+ * @tparam Container The type of container
+ * @tparam Types The template parameters the container accepts (e.g. for a vector it will be the element type
+ *               and the allocator type)
+ * @tparam T The type of the object to push into the container.
+ * @tparam TArgs The type of the remaining objects to push onto the vector (this should match T)
+ * @param container The container to push objects into.
+ * @param first The object to push into the container.
+ * @param args The rest of the objects to push into the container.
  */
-template <typename T, typename...Args>
-void pusher(vector<T>& vec, T first, Args... args) {
-    vec.push_back(first);
-    pusher(vec, args...);
+template <template <typename... Args> class Container, typename... Types, typename T, typename... TArgs>
+void pusher(Container<Types...>& container, T first, TArgs... args) {
```

Comment:
I think that would be easier to understand:
```
template <class CONTAINER, class D, class T, typename... Targs>
void pusher(CONTAINER<D>& cont, T first, Targs... args) {
   ...
}
```
This will make the template specification easier to understand, and it would also allow a flexibility in the `T -> D` conversion.

If the only reason for such complex template is in supporting multi-parametric containers, such as `map<K,V>` then I would rather go with the template specialization by adding extra set of methods for this set of problems. For example:
```
template <class CONTAINER, class K, class V,  class T, typename... Targs>
void pusher(CONTAINER<K,V>& cont, T first, Targs... args) {
   ...
}
```
Since these functions are rather small this won't blow your code. But it would add some clarity to it.